### PR TITLE
gc: Fix GC scan effectiveness to avoid OOM

### PR DESCRIPTION
**Pull request by Connor1996**
Signed-off-by: Connor1996 <zbk602423539@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/11410 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed: limit the pending task for gc worker to avoid OOM

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual Test

Typ8ANc8Hj

S3BBgMGMUL

### Release note 

```release-note
Fix GC scan effectiveness to avoid OOM
```

### CODE REVIEW

File: src/server/gc_worker/compaction_filter.rs [hunk 1]
```diff
@@ -62,9 +62,10 @@ lazy_static! {
     )
     .unwrap();
     // A counter for errors met by `WriteCompactionFilter`.
-    static ref GC_COMPACTION_FAILURE: IntCounter = register_int_counter!(
+    static ref GC_COMPACTION_FAILURE: IntCounterVec = register_int_counter_vec!(
```
Comment (MyonKeminta) [inline src/server/gc_worker/compaction_filter.rs:65]: Does the grafana dashboard need to be changed?

**Review by MyonKeminta — APPROVED** [review]
Rest LGTM

File: src/server/gc_worker/compaction_filter.rs (hunk 1 above)
Comment (BusyJay) [reply src/server/gc_worker/compaction_filter.rs:65]: Is it safe to add a dimension to the existing metrics?

File: src/server/gc_worker/compaction_filter.rs [hunk 2]
```diff
@@ -315,9 +316,16 @@ impl WriteCompactionFilter {
...
+        match self.gc_scheduler.schedule(task) {
+            Ok(_) => {}
+            Err(ScheduleError::Full(_)) => {
+                GC_COMPACTION_FAILURE.with_label_values(&["full"]).inc();
+            }
+            Err(e) => {
+                if log_on_error {
+                    error!("compaction filter schedule {} fail", e);
+                }
+                GC_COMPACTION_FAILURE.with_label_values(&["schedule"]).inc();
```
Comment (BusyJay) [inline src/server/gc_worker/compaction_filter.rs:328]: It can only be disconnected.

File: src/server/gc_worker/compaction_filter.rs [hunk 3]
```diff
@@ -566,7 +574,7 @@ impl CompactionFilter for WriteCompactionFilter {
             Ok(decision) => decision,
             Err(e) => {
                 warn!("compaction filter meet error: {}", e);
-                GC_COMPACTION_FAILURE.inc();
+                GC_COMPACTION_FAILURE.with_label_values(&["err"]).inc();
```
Comment (BusyJay) [inline src/server/gc_worker/compaction_filter.rs:577]: Can it be more specific? We all know it's an error as its name suggested.

File: src/server/gc_worker/gc_worker.rs [hunk 4]
```diff
@@ -782,7 +783,7 @@ where
         cfg: GcConfig,
         feature_gate: FeatureGate,
     ) -> GcWorker<E, RR> {
-        let worker_builder = WorkerBuilder::new("gc-worker");
+        let worker_builder = WorkerBuilder::new("gc-worker").pending_capacity(GC_MAX_PENDING_TASKS);
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:786]: Is compaction filter the only source of the messages?

File: src/server/gc_worker/compaction_filter.rs (hunk 1 above)
Comment (Connor1996) [reply src/server/gc_worker/compaction_filter.rs:65]: "Is it safe to add a dimension to the existing metrics?" changed it in my test cluster, doesn't encounter any problem.

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -16,6 +17,15 @@ pub struct GcInfo {
     pub is_completed: bool,
 }
 
+impl GcInfo {
```
Comment (BusyJay) [inline src/storage/mvcc/txn.rs:20]: Why move?

File: src/server/gc_worker/gc_worker.rs [hunk 6]
```diff
@@ -245,7 +245,8 @@ where
         gc_info.found_versions += next_gc_info.found_versions;
         gc_info.deleted_versions += next_gc_info.deleted_versions;
         gc_info.is_completed = next_gc_info.is_completed;
-        self.stats.add(&reader.statistics);
+        let mut stats = mem::take(&mut reader.statistics);
+        self.stats.add(&mut stats);
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:249]: Why mut?

File: src/storage/mvcc/txn.rs (hunk 5 above)
Comment (Connor1996) [reply src/storage/mvcc/txn.rs:20]: move to where `GcInfo` is defined

File: src/server/gc_worker/gc_worker.rs (hunk 4 above)
Comment (Connor1996) [reply src/server/gc_worker/gc_worker.rs]: for other messages like `GCTask::GC`, `GCTask::UnsafeDestroyRange` and `GCTask::PhysicalScanLock`, they are handled by `handle_gc_task_schedule_error` when encountering error.

**Review by BusyJay — APPROVED** [review]

**Comment by tabokie** [comment]
@BusyJay Turns out it's already picked in #11419, just not labeled.

**Pull request merged**